File: leetcode/0418-sentence-screen-fitting/solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int wordsTyping(vector<string>& sentence, int rows, int cols) {
        string s;
        for (auto& w : sentence) {
            s += w;
            s += ' ';
        }
        int totalLen = s.size();
        long long start = 0;

        for (int i = 0; i < rows; i++) {
            start += cols;
            if (s[start % totalLen] == ' ') {
                start++;
            } else {
                while (start > 0 && s[(start - 1) % totalLen] != ' ') {
                    start--;
                }
            }
        }

        return start / totalLen;
    }
};

File: leetcode/0418-sentence-screen-fitting/solution.py
from typing import List


class Solution:
    def wordsTyping(self, sentence: List[str], rows: int, cols: int) -> int:
        s = ' '.join(sentence) + ' '
        total_len = len(s)
        start = 0

        for _ in range(rows):
            start += cols
            if s[start % total_len] == ' ':
                start += 1
            else:
                while start > 0 and s[(start - 1) % total_len] != ' ':
                    start -= 1

        return start // total_len
